src/compiler/checker.ts

```diff
@@ -2574,6 +2574,40 @@ namespace ts {
             return getMergedSymbol(symbol.parent && getLateBoundSymbol(symbol.parent));
         }
 
+        function getAlternativeContainingModules(symbol: Symbol, enclosingDeclaration: Node): Symbol[] | undefined {
+            const containingFile = getSourceFileOfNode(enclosingDeclaration);
+            if (containingFile && containingFile.imports) {
+                // Try to make an import using an import already in the enclosing file, if possible
+                let results: Symbol[] | undefined;
+                for (const importRef of containingFile.imports) {
+                    if (nodeIsSynthesized(importRef)) continue; // Synthetic names can't be resolved by `resolveExternalModuleName` - they'll cause a debug assert if they error
+                    const resolvedModule = resolveExternalModuleName(enclosingDeclaration, importRef);
+                    if (!resolvedModule) continue;
+                    const ref = getAliasForSymbolInContainer(resolvedModule, symbol);
+                    if (!ref) continue;
+                    results = append(results, resolvedModule);
+                }
+                if (results) {
+                    return results;
+                }
+            }
+            const links = getSymbolLinks(symbol);
+            if (links.extendedContainers) {

```

Comment:
How often will this be called? Does it make sense to cache the above results to prevent recomputing this as well?

Reply:
In `getContainersOfSymbol` which is called by `isAnySymbolAccessible` and thus `isSymbolAccessible` - it's called for every symbol in a program during declaration emit, and some symbols multiple times, at least. May as well cache it - `resolveExternalModuleName` plus `getAliasForSymbolInContainer` is a fair bit of work (just less than doing `getAliasForSymbolInContainer` on every source file in the program).

Reply:
Worth noting though, this needs a separate cache for each containing file.